torchtune/modules/position_embeddings.py

```diff
@@ -125,3 +120,146 @@ def forward(
         # tensor has shape [b, s, n_h, h_d]
         x_out = x_out.flatten(3)
         return x_out.type_as(x)
+
+
+class VisionRotaryPositionalEmbeddings(nn.Module):
+    """
+    This class implements two-dimensional Rotary Positional Embeddings (RoPE) for images
+    based on the axial frequency 2D RoPE described in https://arxiv.org/pdf/2403.13298.
+
+    The position embedding is simply applied to the x-axis and y-axis separately.
+
+    Note: This module assumes the CLS token embedding is appended at the end of the sequence.
+
+    Args:
+        patch_size (int): The size of each patch. Used to divide the tiles into patches.
+            E.g. for ``patch_size=40``, a tile of shape (400, 400) will have 10x10 grid of patches.
+        tile_size (int): The size of your image tiles, if the image was tile-cropped in advance. Otherwise,
+            the size of the input image. In this case, the function will consider your image as a single tile.
+            with shape (40, 40) each.
```

Comment:
a single tile with shape (400, 400)? if max_num_tiles=1, then tile size should be equal to patch_size?

Reply:
ooh ya good catch, let me update this and the docstring where I copied this from

Reply:
I think that "your image as a single tile. with shape (40, 40) each." ---> "your image as a single tile with shape (400, 400)."

I didnt read the code, just sharing what i know from llama 3.2 vision. This explains it visually: https://github.com/pytorch/torchtune/blob/bd89bb88b6092de74e79356ddb193b9e4b614d46/torchtune/modules/vision_transformer.py#L61

A image is broken into tiles. E.g, if max_num_tiles=4 and tile_size=400, a 1600x1600 image is broken into 4 tiles of 400x400, e.g. (num images, num tiles, x, y) = (1,4,400x400)

Then each tile is broken down into patches.

If max_num_tiles=1, then it means that 1 image = 1 tile. The tile dimension doesnt go away, e.g. (num images, num tiles, x, y) = (1,1,x, y)